packages/inferno-mobx/src/observerWrap.ts

```diff
@@ -0,0 +1,116 @@
+import { InfernoNode, render, VNode, createComponentVNode, _HI as normalizeRoot } from 'inferno';
+import { Reaction } from 'mobx';
+import { throwError, warning } from 'inferno-shared';
+import { VNodeFlags } from 'inferno-vnode-flags';
+
+type Render = (properties?: any, context?: Record<string, unknown>) => InfernoNode | undefined | void;
+
+function callDispose({ dispose }: { readonly dispose: () => void }): void {
+  dispose();
+}
+
+interface InnerProperties {
+  readonly context: unknown
+  readonly dispose: () => void
+  readonly props: unknown
+  readonly self: VNode
+  readonly track: (f: () => void) => void
+}
+
+function innerVNode<T>(type: (p: InnerProperties) => T, properties: InnerProperties): VNode {
+  const ref = {
+    onComponentDidUpdate: callDispose,
+    onComponentWillUnmount: properties.dispose
+  };
+  if (process.env.NODE_ENV !== 'production') {
+    Object.freeze(properties);
+    Object.freeze(ref);
+  }
+  return createComponentVNode(VNodeFlags.ComponentFunction, type, properties, undefined, ref);
+}
+
+function makeProxy(target: VNode): { $V: InfernoNode } {
+  return {
+    get $V() {
+      return target.children
+    },
+    set $V(value) {
+      target.children = value
+    }
+  };
+}
+
+export function observerWrap<T extends Render>(base: T): typeof base {
+  if (process.env.NODE_ENV !== 'production') {
+    if (typeof base !== 'function') {
+      throwError(`observerWrap requires a function to wrap, got ${typeof base} instead`);
+    }
+    // @ts-ignore
+    if (base.prototype && base.prototype.render) {
+      throwError("observerWrap should not be applied to constructors.");
+    }
+    // @ts-ignore
+    if (base.isMobXInfernoObserver) {
+      warning("'observerWrap' was used on a component that already has 'observerWrap' applied. Please only apply once");
+    }
+  }
+  function tracked({ context, props, self, track }: InnerProperties): ReturnType<typeof base> {
+    let result;
+    let caught;
+    track(() => {
+      try {
+        result = base.call(self, props, context);
+      } catch(error) {
+        caught = error;
+      }
+    });
+    if (caught) {
+      throw caught;
+    }
+    return result;
+  }
+  function wrapper(this: VNode, props, context): VNode {
+    let onComponentDidUpdate;
```

Comment:
Is this wrapper now the method that will be called on each render ?

Reply:
Yes, for VNodes constructed using the returned function. The goal is that the `wrapper` is a drop in replacement for the `base` function when it comes to making VNodes. A replacement that re-renders itself in response to changes in MobX observables.

The `wrapper` will add more overhead compared to the class components from `observerPatch` when re-rendered by Inferno. Re-rendering triggered by the MobX reaction should be faster though.

In cases where the overhead is a problem, it can be alleviated by adding an `onComponentShouldUpdate` hook to prevent re-rendering if the `props` are unchanged.

The MobX reaction is tied to an 'inner' VNode which the `wrapper` returns. The 'inner' VNode is used to keep the VDOM and real DOM in sync when the reaction performs updates. The 'inner' VNode also has the hooks to dispose of the reaction when the 'inner' VNode is removed either due to being unmounted or because the 'inner' VNode was replaced.